crates/sui-framework/src/natives/crypto/groth16.rs

```diff
@@ -14,22 +13,37 @@ use smallvec::smallvec;
 use std::collections::VecDeque;
 
 pub const INVALID_VERIFYING_KEY: u64 = 0;
+pub const INVALID_CURVE: u64 = 1;
+
+pub const BLS12_831: u8 = 0;
+pub const BN254: u8 = 1;
 
 pub fn prepare_verifying_key(
     _context: &mut NativeContext,
     ty_args: Vec<Type>,
     mut args: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
     debug_assert!(ty_args.is_empty());
-    debug_assert!(args.len() == 1);
+    debug_assert!(args.len() == 2);
 
     let bytes = pop_arg!(args, VectorRef);
     let verifying_key = bytes.as_bytes_ref();
 
+    let curve = pop_arg!(args, u8);
+
     // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
     let cost = legacy_empty_cost();
 
-    match prepare_pvk_bytes(&verifying_key) {
+    let result;
+    if curve == BLS12_831 {
```

Comment:
let result = match... ?

Reply:
Since we need to return an error on invalid curve, it'll be hard to use match in a way that is more readable than this imo.